[10][ADD] Enhance relational fields dropdown and tags

MicheleZak opened the pull request:
This module allow to display additional fields to the many2one dropdown and the many2many tags.

MicheleZak committed:
[10][ADD] Enhance relational fields dropdown and tags

MeetKD commented on web_relational_enhancer/__manifest__.py:
```diff
@@ -0,0 +1,35 @@
+                       "display_fields": "name,state,description"
+                   }'/>
+    """,
+    'author': "Michele Zaccheddu - LevelPrime Srl",
```
Add OCA as a author.

MeetKD commented on web_relational_enhancer/__manifest__.py:
```diff
@@ -0,0 +1,35 @@
+                   }'/>
+    """,
+    'author': "Michele Zaccheddu - LevelPrime Srl",
+    'website': "http://www.levelprime.it",
```
'website': 'https://github.com/OCA/web',

MeetKD commented on web_relational_enhancer/__manifest__.py:
```diff
@@ -0,0 +1,35 @@
+    'author': "Michele Zaccheddu - LevelPrime Srl",
+    'website': "http://www.levelprime.it",
+    'category': 'Web',
+    'version': '1.0',
```
Can you please add proper version number as mentioned in oca guidelines ?

https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

MeetKD commented on web_relational_enhancer/__manifest__.py:
```diff
@@ -0,0 +1,35 @@
+    'website': "http://www.levelprime.it",
+    'category': 'Web',
+    'version': '1.0',
+    'depends': ['base'],
```
Module should be depend on web module.

simahawk commented on web_relational_enhancer/README.rst:
```diff
@@ -0,0 +1,99 @@
+=======================
+
+
+This module allow to add information to the 'display_name' on
```
hmm is this whole machinery really required just to change the display name?
I mean, you can use `context="{'use_special_name: True}"` on the field declaration and customize display name/ name_get to achieve the same result.

Can you elaborate on the use case and the benefit of maintaining this big piece of JS? Tnx :)

MicheleZak replied:
Yes, because it changes the display name only on the dropdown widget, leaving the return values of those methods untouched.

simahawk replied:
this is exactly what I'm talking about :)

you can set the ctx on the field itself and this will enable name_get override (which is used to show the label on the dropdown) only on this field an only when you set the ctx key...

I'm still looking for an answer then :)

MicheleZak replied:
Yeah, ok. I didn't understand what you meant on the first answer. I thought it was a known feature instead of a keyword which you check on the name_get override. 

Your solution is better on a dynamic use case, but I still think it could be good also to have a static way to do that. It is simpler and faster from the usage point of view, because you only have to specify the fields, you want to see in the label, inside the view field options.

simahawk replied:
I still don't agree... you can achieve what you want like:

```xml
context="{'display_fields': ['a', 'b','c']}"
```
then in this module you can simply have the name_get override and use those fields in sequence.
Or  you can even support a format pattern:
```xml
context="{'display_fields': ['a', 'b','c'], 'display_name_pattern': '{c}: {a} - {b}'}"
```
and you are done. This is the only thing you have to do as a user to enjoy the feature.

Just provide the base logic in this module (or a new one, like "web_display_name_improved" or similar...) to handle this context data.
You'll have a few python lines to maintain instead of this huge JS.

Using ctx keys to manipulate/customize look and feel of forms, fields and such is a common and well-know pattern in Odoo world ;)

MicheleZak replied:
You're right. I already used it, I don't know why I did not think about it this time. I'll do it asap.

Thanks for the support.

tarteo commented on web_relational_enhancer/static/src/js/field_enhancer.js:
```diff
@@ -0,0 +1,309 @@
+                    _super.bind(self)(res)
+                })
+            } else
+                this._super(data)
```
Here the variable _super can also be used

tarteo commented on web_relational_enhancer/static/src/js/field_enhancer.js:
```diff
@@ -0,0 +1,309 @@
+                def.resolve(res)
+            })
+        } catch(err) {
+            console.log(err)
```
Can you use console.error here please?

tarteo commented on web_relational_enhancer/static/src/js/field_enhancer.js:
```diff
@@ -0,0 +1,309 @@
+            }).fail(err => console.log(`Error getting relational fields. Error: ${err}`))
+
+            return def
+        } else
```
Else statement is irrelevant here

tarteo commented on web_relational_enhancer/static/src/js/field_enhancer.js:
```diff
@@ -0,0 +1,309 @@
+                    if (item.id) {
+                        self.display_value = {}
+                        self.display_value_backup = {}
+                        self.display_value['' + item.id] = item.name
```
Can u use toString here?

tarteo commented on web_relational_enhancer/static/src/js/field_enhancer.js:
```diff
@@ -0,0 +1,309 @@
+            return data
+    }
+
+    function enhanceMany2OneEntryLabel(result) {
```
Can you move the functions to the field? This will enable other developers to change the behaviors in other modules.

tarteo commented on web_relational_enhancer/static/src/js/field_enhancer.js:
```diff
@@ -0,0 +1,309 @@
+
+    function enhanceMany2OneEntryLabel(result) {
+
+        let self = this
```
Missing semicolon

tarteo commented on web_relational_enhancer/static/src/js/field_enhancer.js:
```diff
@@ -0,0 +1,309 @@
+            this.$input = this.$('input')
+
+            this.init_error_displayer()
+            self.$input.on('focus', function() {
```
You can use `this` here

tarteo commented on web_relational_enhancer/static/src/js/field_enhancer.js:
```diff
@@ -0,0 +1,309 @@
+               this method adds them in the displayed dropdown entry.
+               We can't use super 'cause of this arch
+             */
+            var self = this
```
Why not use let instead of var here?

MicheleZak committed:
[10.0] web_relational_enhancer: Update manifest

- Add OCA as author
- Set correct repo URL
- Set proper version number
- Add 'web' dependency

MicheleZak committed:
[10.0] web_relational_enhancer: Remove irrelevant else, use _super var

- Remove else branch for a return value because irrelevant
- Use '_super' var instead of 'this._super' to call the super method

MicheleZak committed:
[10.0] web_relational_enhancer: Use console messages properly

Switch from 'console.log' method to 'console.warn' and 'console.error'
to report issues on console properly.

MicheleZak committed:
[10.0] web_relational_enhancer: Fix misuse of _super call without context
